pkg/controller/acmechallenges/sync.go

```diff
@@ -72,6 +74,19 @@ func (c *controller) Sync(ctx context.Context, ch *cmacme.Challenge) (err error)
 		return c.handleFinalizer(ctx, ch)
 	}
 
+	// if dns01-check-recreate-period is enabled and the pending challenge's creation time
+	// is more than DNS01CheckRecreatePeriod ealier from now delete the stale challenge so it can be recreated
+	rp := c.DNS01CheckRecreatePeriod
+	if rp > 0 && ch.Spec.Type == v1.ACMEChallengeTypeDNS01 && ch.Status.State == v1.Pending &&
+		ch.CreationTimestamp.Add(rp).Before(time.Now()) {
+		if err := c.cmClient.AcmeV1().Challenges(ch.Namespace).Delete(ctx, ch.Name, metav1.DeleteOptions{}); err != nil {
+			log.Error(err, fmt.Sprintf("error deleting stale challenge %s/%s", ch.Namespace, ch.Name))
+			return err
+		}
+		log.Info(fmt.Sprintf("deleted challenge which lasted longer than the re-create period %s/%s", ch.Namespace, ch.Name))
+		return nil
+	}
```

Comment:
How do we ensure that this code will actually run after the `DNS01CheckRecreatePeriod`?
There may not be any event that would trigger the reconciler at that time.
Would it perhaps make sense to integrate this with the [challenges scheduler](https://github.com/cert-manager/cert-manager/blob/master/pkg/controller/acmechallenges/scheduler/scheduler.go) and make that delete challenges?
(I've not thought about this in depth, so there may be reasons why to not do that)

Reply:
Good point, I think it will eventually run due to [requeue](https://github.com/cert-manager/cert-manager/blob/2c254545d762f53b265ee45d5c0e80b0d17bfa1d/pkg/controller/acmechallenges/sync.go), but not on the exact time when the DNS01CheckRecreatePeriod is reached.

Reply:
[This line](https://github.com/cert-manager/cert-manager/blob/2273662f9481eb38f4a0e3f7d9083dc642c74939/pkg/controller/acmechallenges/sync.go#L210) - is important, right? If the check fails, we requeue after `DNS01CheckRetryPeriod`.

So we could move the check [into this if block, before the requeue](https://github.com/cert-manager/cert-manager/blob/2273662f9481eb38f4a0e3f7d9083dc642c74939/pkg/controller/acmechallenges/sync.go#L200-L213)?

Reply:
> [This line](https://github.com/cert-manager/cert-manager/blob/2273662f9481eb38f4a0e3f7d9083dc642c74939/pkg/controller/acmechallenges/sync.go#L210) - is important, right? If the check fails, we requeue after DNS01CheckRetryPeriod

But that bit only gets to run for 'Processing' challenges as else we return [here](https://github.com/cert-manager/cert-manager/blob/2273662f9481eb38f4a0e3f7d9083dc642c74939/pkg/controller/acmechallenges/sync.go#L100-L104). In this PR we assume that the `Challenge` is 'Pending'

Reply:
I think that make sense, I moved it to the new location before requeue.

Reply:
@irbekrm here is a status I grabbed from a "pending" `Challenge`.
```
status:
  presented: false
  processing: true
  reason: '...'
  state: pending
```

The `Status.Processing` is `true`, but `Status.state` is `pending`.

Reply:
> Status.Processing is true, but Status.State is pending and that's what I was checking.

Thanks for clarifying, that makes sense to me 👍🏼 Apologies I mixed up `status` subfield with the 'state' value!

Reply:
>I'd be in favour of leaving the logic where it is now.

I share the same view.

>Could we perhaps reword the flag description to make this more clear

Absolutely, pushed a change for rewording.